# PipelineProject_Natalie_Stegman

Hello everyone! This is track 2 of the Pipeline Project.

These are the tools that must be installed before running this pipeline:
  - SRA Tool Kit
  - Bowtie2
  - Spades
  - Linux/Unix
  - Python3

# Before you run:
  - Included in this repo are the 8 truncated fastq files
  - You must download these files to run the test data
  - These reads include the first 100000 reads of each read

# Command to run:
**git clone insert link here**
- after you download the truncated dataset and did git clone, you can now run the pipeline:
            **python3 pipeline_project.py**
  
# The files in the produced file include:
  - Trimmed_fastq_files: These are the truncated fastq files so that the pipeline runs under 2 minutes
  - spades_raw_assembly: This is the 
  - mapped_strains:
  - bowtie_indexes:
  - blast: Includes the database of the Betaherpesvirinae family, and the
  - aligned_reads:
  - aligned_reads_unzipped:
  - PipelineProject.log:
  - bowtie_fasta:
